lib/BloqadeKrylov/src/cfet.jl

```diff
@@ -0,0 +1,205 @@
+
+"""
+    CFET42Evolution(reg, start_clock, durations, hamiltonian, options)
+
+Create a `CFET42Evolution` object.
+
+# Arguments
+
+- `reg`: a register object.
+- `start_clock`: start clock of the evolution.
+- `durations`: list of durations at each time step.
+- `hamiltonian`: low-level hamiltonian object of type [`Hamiltonian`](@ref).
+- `options`: options of the evolution in type [`KrylovOptions`](@ref).
+"""
+function CFET42Evolution(reg, start_clock, durations, hamiltonian, options)
+    return CFET42Evolution{typeof(reg),typeof(start_clock),typeof(hamiltonian)}(
+        reg,
+        start_clock,
+        durations,
+        hamiltonian,
+        options,
+    )
+end
+
+function Adapt.adapt_structure(to, x::CFET42Evolution)
+    return CFET42Evolution(adapt(to, x.reg), x.start_clock, x.durations, adapt(to, x.hamiltonian), x.options)
+end
+
+function CFET42Evolution(reg::AbstractRegister, clocks, h; kw...)
+    all(≥(0), clocks) || throw(ArgumentError("clocks must not be negative"))
+    options = from_kwargs(KrylovOptions; kw...)
+    P = real(eltype(statevec(reg)))
+    T = isreal(h) ? P : Complex{P}
+    start_clock, durations = first(clocks), diff(clocks)
+    
+    ## checking if it is equal time:
+    if length(unique(round.(durations,digits = 14) ) ) != 1
+        throw(ArgumentError("durations must be equal (time slice must be equal)"))
+    end 
+
+    return CFET42Evolution(reg, start_clock, durations, Hamiltonian(T, h, space(reg)), options)
+end
+
+
+#=
+# this is the naive implementation of CFET42Evolution
+function emulate_step!(prob::CFET42Evolution, step::Int, clock::Real, duration::Real)
+    state = statevec(prob.reg)
+    h = prob.hamiltonian
+
+    A1 = BloqadeExpr.to_matrix(h(clock + (0.5-√3/6)*duration))
+    A2 = BloqadeExpr.to_matrix(h(clock + (0.5+√3/6)*duration)) 
+    
+    s1 = (3 + 2*√3)/12
+    s2 = (3 - 2*√3)/12
+
+    # exp stage 1
+    Ω4 = s1*A1 + s2*A2
+    expmv!(-im*duration, Ω4, state; prob.options.tol) 
+
+    # exp stage 2
+    Ω4 = s2*A1 + s1*A2
+    expmv!(-im*duration, Ω4, state; prob.options.tol) 
+
+    # do we need this normalization? 
+    if mod(step, prob.options.normalize_step) == 0
+        normalize!(prob.reg)
+    end
+
+    if prob.options.normalize_finally && step == length(prob.durations)
+        normalize!(prob.reg)
+    end
+    
+    return prob
+end
+=#
+
+# optimize by putting addition of two hamiltonians 
+# into the fs(t) coefficients at time t1 & t2
+function emulate_step!(prob::CFET42Evolution, step::Int, clock::Real, duration::Real)
+
+    state = statevec(prob.reg)
+    h = prob.hamiltonian
+
+    t1 = clock + (0.5-√3/6)*duration
+    t2 = clock + (0.5+√3/6)*duration
+    
+    #for i in range(size(A1.h.fs)):
+    a1 = collect(map(h.fs) do f 
+                    return f(t1)
+                end
+                )
+    a2 = collect(map(h.fs) do f 
+                    return f(t2)
+                end
+                )
+
+    s1 = (3 + 2*√3)/12
+    s2 = (3 - 2*√3)/12
```

Comment:
Can you create some kind of abstraction to enable higher-order solvers of this sort in the future?

Reply:
Yeah if we do the approach I mentioned (TIHamiltonian), then this part will become even human readable because we can simply make A1 and A2 a TIHamiltonian, then expression will be just like Magnus4